### PR TITLE
Add disjoint-sets feature to fipmapper

### PR DESCRIPTION
**Pull request by berland**
Solves the problem in https://equinor.slack.com/archives/C3HLP8ZTQ/p1620809891326900

### CODE REVIEW

**Comment by berland** [comment]
@tnatt this is for your "Source comparison" webviz volumetric viewer.

File: tests/fipmapper/test_disjointsets.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+                {"REGION": "A", "ZONE": "L", "FIPNUM": 1, "ROOT": ("A", "U", 1)},
+                {"REGION": "A", "ZONE": "U", "FIPNUM": 1, "ROOT": ("A", "U", 1)},
+                {"REGION": "B", "ZONE": "L", "FIPNUM": 2, "ROOT": ("B", "U", 2)},
+                {"REGION": "B", "ZONE": "U", "FIPNUM": 2, "ROOT": ("B", "U", 2)},
+            ],
+        ),
+        (
+            # Almost three-by-three reservoir, used in demonstration
+            # of the problem solved, involving all "cruxes":
+            {
```
Comment (markusdregi) [inline tests/fipmapper/test_disjointsets.py:102]: I need a bit more context to be able to review this. In particular, I tried to understand the tests by inspecting them and the documentation, but did not fully get it. Could you provide a more in-details explanation of the input and output of this example? :)

File: src/fmu/tools/fipmapper/fipmapper.py [hunk 2]
```diff
@@ -167,6 +171,21 @@ def _fipdata_from_webvizyaml(self, yamldict: dict):
     def _fips2regions(self, fips: List[int]) -> List[List[str]]:
         return [self.fip2region(fip_int) for fip_int in fips]
 
+    def get_regions(self) -> List[str]:
```
Comment (markusdregi) [inline src/fmu/tools/fipmapper/fipmapper.py:174]: These should all get their own tests with both happy and not-so-happy path being tested...

Comment (berland) [reply src/fmu/tools/fipmapper/fipmapper.py:174]: Fixed.

File: tests/fipmapper/test_disjointsets.py (hunk 1 above)
Comment (berland) [reply tests/fipmapper/test_disjointsets.py:182]: ASCII-art added, and the example is elevated to be the first test. Hopefully it helps..

File: src/fmu/tools/fipmapper/fipmapper.py [hunk 3]
```diff
@@ -270,6 +289,112 @@ def regzone2fip(self, region: str, zone: str) -> List[int]:
...
+                x["REGION_y"],
+                x["ZONE_y"],
+                x["FIPNUM_y"],
+            ),
+            axis=1,
+        )
+        # Filter to only edges that determine which cell linkages
+        # that should be grouped:
+        neighbourlist = edges[edges["NEIGHBOURS"]]
+        print(neighbourlist)
```
[the remark this answers is not in the document]
Comment (berland) [reply src/fmu/tools/fipmapper/fipmapper.py:355]: ouch. Fixed.

File: src/fmu/tools/fipmapper/fipmapper.py [hunk 4]
```diff
@@ -270,6 +289,112 @@ def regzone2fip(self, region: str, zone: str) -> List[int]:
...
+    A pair of reg-zone-fip is in the same group if they must be treated together
+    when (and have properties summed). Say if you have a value for
+    a specific region, but this region contains two FIPNUMs. Then we can never
+    treat these two FIPNUMs separately, they must be summed in order to be
+    comparable to the value for the region
+    """
+    if (reg1 == reg2) and (zon1 == zon2) and (fip1 == fip2):
+        # The mapping to itself is to be ignored.
+        return False
+    if ((reg1 == reg2) and (zon1 == zon2)) or (fip1 == fip2):
```
[the remark this answers is not in the document]
Comment (berland) [reply src/fmu/tools/fipmapper/fipmapper.py:394]: Right! The mapping to itself can be truly ignored..

**Comment by berland** [comment]
> I would consider changing the type of the root in the data frame though. As although a cell within the partition identifies it uniquely, it is easy to put more into it then what it really is... I would map it to integers instead as a last step, but I leave the decision for you :)

Agree, now changed to integers, and lookup functions based on these integers added.

**Pull request merged**